src/appleseed/renderer/kernel/rendering/baserenderer.cpp

```diff
@@ -0,0 +1,225 @@
+
+    m_shading_system->attribute(
+        "raytypes",
+        OSL::TypeDesc(
+            OSL::TypeDesc::STRING,
+            static_cast<int>(VisibilityFlags::Count)),
+        VisibilityFlags::Names);
+
+#ifndef NDEBUG
+    // While debugging, we want all possible outputs.
+    m_shading_system->attribute("debug", 1);
+    m_shading_system->attribute("compile_report", 1);
+    m_shading_system->attribute("countlayerexecs", 1);
+    m_shading_system->attribute("clearmemory", 1);
+#endif
+
+    // Register appleseed's closures into OSL's shading system.
+    register_closures(*m_shading_system);
+#endif
+}
+
+BaseRenderer::~BaseRenderer()
+{
+#ifdef APPLESEED_WITH_OSL
+    RENDERER_LOG_DEBUG("%s.", "releasing optimized OSL shader groups");
+    m_project.get_scene()->release_optimized_osl_shader_groups();
+
+    RENDERER_LOG_DEBUG("%s.", "destroying OSL shading system");
+    OSL::ShadingSystem::destroy(m_shading_system);
+
+    RENDERER_LOG_DEBUG("%s.", "destroying OSL renderer services");
+    delete m_renderer_services;
+#endif
+
+#ifdef APPLESEED_WITH_OIIO
+    const string stats = m_texture_system->getstats();
+    const string trimmed_stats = trim_right(stats, "\r\n");
+    RENDERER_LOG_INFO("%s", trimmed_stats.c_str());
+
+    RENDERER_LOG_DEBUG("%s.", "destroying OpenImageIO texture system");
+    OIIO::TextureSystem::destroy(m_texture_system);
+
+    RENDERER_LOG_DEBUG("%s.", "destroying OpenImageIO error handler");
+    delete m_error_handler;
+#endif
+}
+
+ParamArray& BaseRenderer::get_parameters()
+{
+    return m_params;
+}
+
+const ParamArray& BaseRenderer::get_parameters() const
+{
+    return m_params;
+}
+
+bool BaseRenderer::initialize_shading_system(
+    TextureStore& texture_store,
+    IAbortSwitch& abort_switch)
+{
+#ifdef APPLESEED_WITH_OIIO
+    const ParamArray& params = m_params.child("texture_store");
+
+    const size_t texture_cache_size_bytes =
+        params.get_optional<size_t>("max_size", 256 * 1024 * 1024);
+
+    RENDERER_LOG_INFO(
+        "setting OpenImageIO texture cache size to %s.",
+        pretty_size(texture_cache_size_bytes).c_str());
+
+    const float texture_cache_size_mb =
+        static_cast<float>(texture_cache_size_bytes) / (1024 * 1024);
+
+    m_texture_system->attribute("max_memory_MB", texture_cache_size_mb);
+
+    // search paths
+    {
+        string prev_search_path;
+        m_texture_system->getattribute("searchpath", prev_search_path);
+        const string new_search_path = m_project.make_search_path_string();
+
+        if (new_search_path != prev_search_path)
+        {
+            RENDERER_LOG_INFO(
+                "setting OpenImageIO search path to %s.",
+                new_search_path.c_str());
+
+            m_texture_system->clear();
+            m_texture_system->attribute("searchpath", new_search_path);
+        }
+    }
+#endif
+
+#ifdef APPLESEED_WITH_OSL
+    m_renderer_services->initialize(texture_store);
+
+    // search paths
```

Comment:
What we get is a much shorter function, no need for artificial scopes, and no need for comments (just name the functions based on the comments you added, e.g. `update_oiio_texture_sys_attrs`).